fdbserver/storageserver.actor.cpp

```diff
@@ -4906,6 +4967,421 @@ ACTOR Future<Void> validateRangeAgainstServers(StorageServer* data,
+				TraceEvent(SevVerbose, "AuditStorageShardStorageServerShardReadRange", serverID)
+				    .detail("Range", range)
+				    .detail("Prefix", serverKeysPrefixFor(serverID))
+				    .detail("ResultSize", readResult.size())
+				    .detail("ReadRangeCount", readRangeCount)
+				    .detail("RetryCount", retryCount)
+				    .detail("AduitServerID", serverID);
+				readRangeCount++;
+				for (int i = 0; i < readResult.size() - 1; ++i) {
+					TraceEvent(SevVerbose, "AuditStorageShardStorageServerShardReadRangeAddToResult", serverID)
+					    .detail("ValueIsServerKeysFalse", readResult[i].value == serverKeysFalse)
+					    .detail("ServerHasKey", serverHasKey(readResult[i].value))
+					    .detail("Range", KeyRangeRef(readResult[i].key, readResult[i + 1].key))
+					    .detail("AuditServer", serverID.first());
+					if (serverHasKey(readResult[i].value)) {
+						res.push_back(Standalone(KeyRangeRef(readResult[i].key, readResult[i + 1].key)));
+					}
+				}
+				if (!readResult.more) {
+					break;
+				}
+				keyBegin = readResult.back().key;
+			}
+			break;
+		} catch (Error& e) {
+			TraceEvent(SevDebug, "AuditStorageShardStorageServerShardeError", serverID)
+			    .errorUnsuppressed(e)
+			    .detail("AuditServer", serverID.first());
+			wait(tr->onError(e));
+			readRangeCount = 0;
+			retryCount++;
+		}
+	}
+	return res;
+}
+
+ACTOR Future<Void> auditStorageStorageServerShardQ(StorageServer* data, AuditStorageRequest req) {
+	TraceEvent(SevInfo, "AuditStorageShardStorageServerShardBegin", data->thisServerID);
+	ASSERT(req.getType() == AuditType::ValidateStorageServerShard);
+	wait(data->serveAuditStorageParallelismLock.take(TaskPriority::DefaultYield));
+	state FlowLock::Releaser holder(data->serveAuditStorageParallelismLock);
+	state Version grv;
+	state Version readAtVersion;
+	state Version versionBeforeWait;
+	state std::vector<std::pair<Version, KeyRangeRef>> shardAssignments;
+	state AuditStorageState res(req.id, data->thisServerID, req.getType());
+	state std::vector<std::string> errors;
+	state Transaction tr(data->cx);
+	// We want to find out any mismatch between ownRangesSeenByServerKey and ownRangesLocalView
+	state std::vector<KeyRange> ownRangesSeenByServerKey;
+	state std::vector<KeyRange> ownRangesLocalView;
+	try {
+		if (data->version.get() == 0) {
+			TraceEvent(SevInfo, "AuditStorageShardStorageServerShardFailed", data->thisServerID)
+			    .detail("Reason", "Version is not valid")
+			    .detail("LatestVersion", data->version.get());
+			req.reply.sendError(audit_storage_failed()); // we cannot run this aduit for this time
+			return Void();
+		}
+
+		// Read local shard info and serverKeys
+		tr.setOption(FDBTransactionOptions::READ_SYSTEM_KEYS);
+		wait(store(grv, tr.getReadVersion()));
+		// When selecting which version to read, we should read at a new version
+		// As a result, any fall behind SS can eventually read the data
+		if (grv <= data->version.get()) {
```

Comment:
Do we need to get the GRV first, can we just issue a read blindly to the `serverKeys` space, and use that read version, i.e., `tr.getReadVersion()` after the read?

Reply:
This algorithm wants to know the current GRV version before read the data, because the algorithm wants to compare current SS version and GRV version to decide which version is good to read.

For local shardInfo read, we always read at data->version.get(). For transaction read, we read at `readAtVersion` version. `readAtVersion` is decided by comparing GRV and data->version.get(). If GRV <= data->version.get(), then readAtVersion is data->version.get(). If GRV > data->version.get(), we wait until data->version.get() is as large as GRV, and readAtVersion is the new data->version.get().